# HW2 on Data Science course of Sharif University of Technology
## Created by: Mohammad Mahdi Hossein Beiky     SI: 400100995
---

### Import python packages here:

In [100]:
import pandas as pd

### First step: Reading dataset
I just need to import the csv file containing achareh dataset and then see the first few rows of it.

In [101]:
achare = pd.read_csv('achare.csv')
achare

No. Previous Orders by Customer                Customer Join Date  \
0                                    3  1396-06-09 11:51:32.505950+04:30   
1                                    0  1400-06-24 02:54:34.549933+04:30   
2                                    0  1400-06-24 02:46:15.891303+04:30   
3                                    0  1397-03-16 14:28:00.764511+04:30   
4                                    0  1400-06-24 02:26:08.327191+04:30   
...                                ...                               ...   
15348                                5  1397-12-20 08:14:12.107241+03:30   
15349                                2  1399-03-12 09:02:46.650380+04:30   
15350                                2  1399-03-11 11:54:27.008224+04:30   
15351                                7  1398-09-04 19:35:23.226869+03:30   
15352                               10  1397-08-14 19:01:50.168271+03:30   

       Parent Service Category               Service Name   Order Status  \
0              نظافت و پذیرایی             نظافت راه‌ پله       ثبت‌ شده   
1          دکوراسیون و بازسازی                      بنایی  Achar_Pending   
2      نصب و تعمیر لوازم خانگی  نصب و تعمیر یخچال و فریزر       ثبت‌ شده   
3                تعمیرات خودرو            صافکاری و نقاشی       ثبت‌ شده   
4          دکوراسیون و بازسازی                      بنایی        لغو شده   
...                        ...                        ...            ...   
15348      دکوراسیون و بازسازی         تعمیر و نصب کابینت       پیدا نشد   
15349      دکوراسیون و بازسازی                      بنایی       پیدا نشد   
15350          نظافت و پذیرایی           خدمات عادی نظافت       پیدا نشد   
15351                    شستشو                   قالیشویی       پیدا نشد   
15352                    آموزش               آموزش موسیقی        لغو شده   

      Order City       Order Region Workman Joine Date  \
0          تهران         استاد معین                NaN   
1          تهران            وردآورد                NaN   
2            رشت          پل طالشان                NaN   
3            رشت          فلکه جهاد                NaN   
4          تهران            ستارخان                NaN   
...          ...                ...                ...   
15348      سنندج  پاسداران (۶ بهمن)                NaN   
15349      سنندج         بلوار شبلی                NaN   
15350      سنندج             بهاران                NaN   
15351      سنندج             بهاران                NaN   
15352      تهران            سهروردی                NaN   

                           Desired Date                     Creation Date  \
0      1400-06-26 10:00:01.000000+04:30  1400-06-24 03:06:56.093221+04:30   
1      1400-06-24 09:30:01.000000+04:30  1400-06-24 03:03:12.365980+04:30   
2      1400-06-25 12:30:01.000000+04:30  1400-06-24 02:59:20.374623+04:30   
3      1400-06-24 09:00:01.000000+04:30  1400-06-24 02:42:10.924776+04:30   
4      1400-06-24 09:30:01.000000+04:30  1400-06-24 02:37:03.244927+04:30   
...                                 ...                               ...   
15348  1400-06-23 09:30:01.000000+04:30  1399-03-12 09:56:30.525000+04:30   
15349  1400-06-23 18:00:02.000000+04:30  1399-03-12 09:05:23.098994+04:30   
15350  1400-06-23 14:00:02.000000+04:30  1399-03-11 21:45:13.938527+04:30   
15351  1400-06-23 11:00:01.000000+04:30  1399-03-11 17:16:58.318792+04:30   
15352  1487-10-10 08:30:01.000000+03:30  1397-10-09 15:30:49.000000+03:30   

                            Finish Date  Customer Rating  \
0      1400-06-26 16:00:01.000000+04:30              NaN   
1      1400-06-24 10:30:01.000000+04:30              NaN   
2      1400-06-25 13:30:01.000000+04:30              NaN   
3      1400-06-24 10:00:01.000000+04:30              NaN   
4      1400-06-24 10:30:01.000000+04:30              NaN   
...                                 ...              ...   
15348  1400-06-23 10:30:01.000000+04:30              NaN   
15349  1400-06-23 19:00:02.000000+04:30              NaN   
15350  1400-06-23 1

In [102]:
serie = achare.isnull().sum()
print(serie)

No. Previous Orders by Customer        0
Customer Join Date                     0
Parent Service Category                0
Service Name                           0
Order Status                           0
Order City                             0
Order Region                           0
Workman Joine Date                  8315
Desired Date                           0
Creation Date                          0
Finish Date                            0
Customer Rating                    13520
workman Rating Score               15353
Comment                            14629
Device                                79
Price                                  0
Discount                               0
Quoted Task?                           0
dtype: int64


### Second step: Handling missing values
#### At the first look we'll get that there are a few columns that have a lot of entries with NaN values
#### These columns are: `Workman Joine Data`, `Customer Rating`, `Workman Rating Score`, `Comment`, and `Device`
#### Let's look at the number of missing entries of each of them and then decide what to do.


In [103]:
workman_join_date_null = achare['Workman Joine Date'].isnull().sum()
customer_rating_null = achare['Customer Rating'].isnull().sum()
workman_rating_score_null = achare['workman Rating Score'].isnull().sum()
comment_null = achare['Comment'].isnull().sum()
device_null = achare['Device'].isnull().sum()
print('Number of NaN values in each column:\nColumn: Workman Joine Date: ', workman_join_date_null
      , '\nColumn: Customer Rating: ', customer_rating_null
      , '\nColumn: Workman Rating Score: ', workman_rating_score_null
      , '\nColumn: Comment: ', comment_null
      , '\nColumn: Device: ', device_null)

Number of NaN values in each column:
Column: Workman Joine Date:  8315 
Column: Customer Rating:  13520 
Column: Workman Rating Score:  15353 
Column: Comment:  14629 
Column: Device:  79


#### We can see that those numbers are huge and in the other hand, the number of dataset rows are: 15353
#### So we can say that these columns must be filled or dropped.
#### The number of first column `Workman Joine Date` null values are the half of its total entries. So for this column we can drop the rows whose this column value is NaN. But for the other three columns, we drop the whole column due to the massiveness of number. Also for the `Device` column there are only 79 empty entries so we just drop the rows corresponding to these entries.

In [104]:
achare = achare.drop(columns=['Customer Rating', 'workman Rating Score', 'Comment'])

In [105]:
achare = achare.dropna(subset = ['Workman Joine Date', 'Device'])


In [106]:
achare

No. Previous Orders by Customer                Customer Join Date  \
18                                   3  1400-03-30 14:22:43.539774+04:30   
20                                   3  1399-05-08 01:01:18.941039+04:30   
23                                   0  1400-06-24 01:18:17.991440+04:30   
24                                  10  1400-05-21 11:55:52.500748+04:30   
27                                   0  1399-09-24 09:01:42.685711+03:30   
...                                ...                               ...   
15338                               26  1398-12-02 22:37:45.592924+03:30   
15339                               26  1398-12-02 22:37:45.592924+03:30   
15340                              112  1396-09-24 09:55:48.408000+03:30   
15341                                1  1399-09-25 15:00:52.436593+03:30   
15342                              199  1396-10-11 11:15:31.507745+03:30   

       Parent Service Category           Service Name Order Status Order City  \
18     نصب و تعمیر لوازم خانگی   نصب و تعمیر مایکروفر  تخصیص یافته      تهران   
20             سرمایش و گرمایش          تعمیر آبگرمکن  تخصیص یافته      تهران   
23                    لوله کشی  لوله‌ کشی آب و فاضلاب  تخصیص یافته        کرج   
24                     برقکاری      سیم کشی و برقکاری  تخصیص یافته      تهران   
27               زیبایی بانوان                براشینگ  تخصیص یافته      تهران   
...                        ...                    ...          ...        ...   
15338          نظافت و پذیرایی   سرویس دوره‌ ای نظافت    اتمام کار      تهران   
15339          نظافت و پذیرایی   سرویس دوره‌ ای نظافت    اتمام کار      تهران   
15340      دکوراسیون و بازسازی        نصب کاغذ دیواری    اتمام کار     اصفهان   
15341          نظافت و پذیرایی       خدمات ویژه نظافت    اتمام کار        رشت   
15342      دکوراسیون و بازسازی        نصب کاغذ دیواری    اتمام کار     اصفهان   

          Order Region                Workman Joine Date  \
18     مصلی امام خمینی  1400-02-02 00:35:51.353573+04:30   
20           یافت آباد  1398-11-28 12:51:26.505459+03:30   
23        کیانمهر(کرج)  1398-10-17 00:23:06.892797+03:30   
24          سعادت آباد  1398-04-15 12:02:22.607681+04:30   
27               نارمک  1398-11-22 18:18:25.328767+03:30   
...                ...                               ...   
15338       استاد معین  1398-10-17 10:00:47.563189+03:30   
15339       استاد معین  1398-10-17 10:00:47.563189+03:30   
15340               جی  1396-09-24 09:55:48.408000+03:30   
15341          قلی پور  1398-08-21 14:51:38.048495+03:30   
15342               جی  1396-09-24 09:55:48.408000+03:30   

                           Desired Date                     Creation Date  \
18     1400-06-24 06:00:02.000000+04:30  1400-06-24 01:36:07.381697+04:30   
20     1400-06-25 09:30:01.000000+04:30  1400-06-24 01:34:12.015376+04:30   
23     1400-06-24 10:00:01.000000+04:30  1400-06-24 01:23:29.522597+04:30   
24     1400-06-24 04:00:02.000000+04:30  1400-06-24 01:20:38.603347+04:30   
27     1400-06-25 09:30:02.000000+04:30  1400-06-24 01:17:37.837603+04:30   
...                                 ...                               ...   
15338  1400-08-08 10:30:02.000000+03:30  1400-03-26 14:14:22.324115+04:30   
15339  1400-07-08 10:30:02.000000+03:30  1400-03-26 14:14:04.752312+04:30   
15340  1400-08-15 13:30:02.000000+03:30  1400-02-02 11:18:22.101495+04:30   
15341  1400-09-27 09:00:02.000000+03:30  1399-09-25 15:01:09.779080+03:30   
15342  1400-09-12 06:00:01.000000+03:30  1399-09-11 11:09:01.498732+03:30   

                            Finish Date   Device  Price  Discount  \
18     1400-06-24 07:00:02.000000+04:30      web      0         0   
20     1400-06-25 10:30:01.000000+04:30  android      0         0   
23     1400-06-24 11:00:01.000000+04:30  android      0         0   
24     1400-06-24 05:00:02.000000+04:30      web      0         0   
27     1400-06-25 10:30:02.000000+04:30      web      0         0   
...                                 ...      ...    ...      

#### So now let's check how many null values are left

In [107]:
serie = achare.isnull().sum()
print(serie)

No. Previous Orders by Customer    0
Customer Join Date                 0
Parent Service Category            0
Service Name                       0
Order Status                       0
Order City                         0
Order Region                       0
Workman Joine Date                 0
Desired Date                       0
Creation Date                      0
Finish Date                        0
Device                             0
Price                              0
Discount                           0
Quoted Task?                       0
dtype: int64


#### Nothing, excellent.
#### Now as the third step, let's look at the name of columns to see what's going on.
### Third step: Renaming the columns

In [108]:
achare.columns

Index(['No. Previous Orders by Customer', 'Customer Join Date',
       'Parent Service Category', 'Service Name', 'Order Status', 'Order City',
       'Order Region', 'Workman Joine Date', 'Desired Date', 'Creation Date',
       'Finish Date', 'Device', 'Price', 'Discount', 'Quoted Task?'],
      dtype='object')

#### For our sake, let's make all the headers' name lowercase, the next step is to correct the name of `Workman Joine Date` column. Then for the last step we remove the question mark in the last column name.

In [113]:
achare.columns = map(str.lower, achare.columns)
achare = achare.rename(columns = {'workman joine date': 'workman join date', 'quoted task?': 'quoted task'})
achare.columns

Index(['no. previous orders by customer', 'customer join date',
       'parent service category', 'service name', 'order status', 'order city',
       'order region', 'workman join date', 'desired date', 'creation date',
       'finish date', 'device', 'price', 'discount', 'quoted task'],
      dtype='object')

#### Now let's get back to the upper case format.

In [120]:
achare.columns = map(str.title, achare.columns)
achare.columns

Index(['No. Previous Orders By Customer', 'Customer Join Date',
       'Parent Service Category', 'Service Name', 'Order Status', 'Order City',
       'Order Region', 'Workman Join Date', 'Desired Date', 'Creation Date',
       'Finish Date', 'Device', 'Price', 'Discount', 'Quoted Task'],
      dtype='object')

#### If you run the code below,
```
achare
```
#### You'll notice that the indices of each row is scrambled, so we need to reset these numbers.

In [132]:
achare.reset_index(drop=True, inplace=True)
achare

No. Previous Orders By Customer                Customer Join Date  \
0                                   3  1400-03-30 14:22:43.539774+04:30   
1                                   3  1399-05-08 01:01:18.941039+04:30   
2                                   0  1400-06-24 01:18:17.991440+04:30   
3                                  10  1400-05-21 11:55:52.500748+04:30   
4                                   0  1399-09-24 09:01:42.685711+03:30   
...                               ...                               ...   
7033                               26  1398-12-02 22:37:45.592924+03:30   
7034                               26  1398-12-02 22:37:45.592924+03:30   
7035                              112  1396-09-24 09:55:48.408000+03:30   
7036                                1  1399-09-25 15:00:52.436593+03:30   
7037                              199  1396-10-11 11:15:31.507745+03:30   

      Parent Service Category           Service Name Order Status Order City  \
0     نصب و تعمیر لوازم خانگی   نصب و تعمیر مایکروفر  تخصیص یافته      تهران   
1             سرمایش و گرمایش          تعمیر آبگرمکن  تخصیص یافته      تهران   
2                    لوله کشی  لوله‌ کشی آب و فاضلاب  تخصیص یافته        کرج   
3                     برقکاری      سیم کشی و برقکاری  تخصیص یافته      تهران   
4               زیبایی بانوان                براشینگ  تخصیص یافته      تهران   
...                       ...                    ...          ...        ...   
7033          نظافت و پذیرایی   سرویس دوره‌ ای نظافت    اتمام کار      تهران   
7034          نظافت و پذیرایی   سرویس دوره‌ ای نظافت    اتمام کار      تهران   
7035      دکوراسیون و بازسازی        نصب کاغذ دیواری    اتمام کار     اصفهان   
7036          نظافت و پذیرایی       خدمات ویژه نظافت    اتمام کار        رشت   
7037      دکوراسیون و بازسازی        نصب کاغذ دیواری    اتمام کار     اصفهان   

         Order Region                 Workman Join Date  \
0     مصلی امام خمینی  1400-02-02 00:35:51.353573+04:30   
1           یافت آباد  1398-11-28 12:51:26.505459+03:30   
2        کیانمهر(کرج)  1398-10-17 00:23:06.892797+03:30   
3          سعادت آباد  1398-04-15 12:02:22.607681+04:30   
4               نارمک  1398-11-22 18:18:25.328767+03:30   
...               ...                               ...   
7033       استاد معین  1398-10-17 10:00:47.563189+03:30   
7034       استاد معین  1398-10-17 10:00:47.563189+03:30   
7035               جی  1396-09-24 09:55:48.408000+03:30   
7036          قلی پور  1398-08-21 14:51:38.048495+03:30   
7037               جی  1396-09-24 09:55:48.408000+03:30   

                          Desired Date                     Creation Date  \
0     1400-06-24 06:00:02.000000+04:30  1400-06-24 01:36:07.381697+04:30   
1     1400-06-25 09:30:01.000000+04:30  1400-06-24 01:34:12.015376+04:30   
2     1400-06-24 10:00:01.000000+04:30  1400-06-24 01:23:29.522597+04:30   
3     1400-06-24 04:00:02.000000+04:30  1400-06-24 01:20:38.603347+04:30   
4     1400-06-25 09:30:02.000000+04:30  1400-06-24 01:17:37.837603+04:30   
...                                ...                               ...   
7033  1400-08-08 10:30:02.000000+03:30  1400-03-26 14:14:22.324115+04:30   
7034  1400-07-08 10:30:02.000000+03:30  1400-03-26 14:14:04.752312+04:30   
7035  1400-08-15 13:30:02.000000+03:30  1400-02-02 11:18:22.101495+04:30   
7036  1400-09-27 09:00:02.000000+03:30  1399-09-25 15:01:09.779080+03:30   
7037  1400-09-12 06:00:01.000000+03:30  1399-09-11 11:09:01.498732+03:30   

                           Finish Date   Device  Price  Discount  \
0     1400-06-24 07:00:02.000000+04:30      web      0         0   
1     1400-06-25 10:30:01.000000+04:30  android      0         0   
2     1400-06-24 11:00:01.000000+04:30  android      0         0   
3     1400-06-24 05:00:02.000000+04:30      web      0         0   
4     1400-06-25 10:30:02.000000+04:30      web      0         0   
...                                ...      ...    ...       ...   
7033  1400-08-08 11:30:02.000000+03:30      we